// kurasuta/dist/Cluster/BaseCluster.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCluster = void 0;
const ShardClientUtil_1 = require("../Sharding/ShardClientUtil");
const Constants_1 = require("../Util/Constants");
const Util = __importStar(require("../Util/Util"));
class BaseCluster {
    constructor(manager) {
        this.manager = manager;
        const env = process.env;
        const shards = env.CLUSTER_SHARDS.split(',').map(Number);
        const clientConfig = Util.mergeDefault(manager.clientOptions, {
            shards,
            shardCount: Number(env.CLUSTER_SHARD_COUNT)
        });
        this.client = new manager.client(clientConfig);
        const client = this.client;
        client.shard = new ShardClientUtil_1.ShardClientUtil(client, manager.ipcSocket);
        this.id = Number(env.CLUSTER_ID);
    }
    async init() {
        const shardUtil = this.client.shard;
        await shardUtil.init();
        this.client.once('ready', () => { void shardUtil.send({ op: Constants_1.IPCEvents.READY, d: this.id }, { receptive: false }); });
        this.client.on('shardReady', id => { void shardUtil.send({ op: Constants_1.IPCEvents.SHARDREADY, d: { id: this.id, shardID: id } }, { receptive: false }); });
        this.client.on('shardReconnecting', id => { void shardUtil.send({ op: Constants_1.IPCEvents.SHARDRECONNECT, d: { id: this.id, shardID: id } }, { receptive: false }); });
        this.client.on('shardResume', (id, replayed) => { void shardUtil.send({ op: Constants_1.IPCEvents.SHARDRESUME, d: { id: this.id, shardID: id, replayed } }, { receptive: false }); });
        this.client.on('shardDisconnect', ({ code, reason, wasClean }, id) => { void shardUtil.send({ op: Constants_1.IPCEvents.SHARDDISCONNECT, d: { id: this.id, shardID: id, closeEvent: { code, reason, wasClean } } }, { receptive: false }); });
        await this.launch();
    }
}
exports.BaseCluster = BaseCluster;
