// kurasuta/dist/Util/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharderEvents = exports.IPCEvents = exports.version = exports.http = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
exports.http = {
    version: 8,
    api: 'https://discordapp.com/api'
};
exports.version = '3.0.1';
var IPCEvents;
(function (IPCEvents) {
    IPCEvents[IPCEvents["EVAL"] = 0] = "EVAL";
    IPCEvents[IPCEvents["MESSAGE"] = 1] = "MESSAGE";
    IPCEvents[IPCEvents["BROADCAST"] = 2] = "BROADCAST";
    IPCEvents[IPCEvents["READY"] = 3] = "READY";
    IPCEvents[IPCEvents["SHARDREADY"] = 4] = "SHARDREADY";
    IPCEvents[IPCEvents["SHARDRECONNECT"] = 5] = "SHARDRECONNECT";
    IPCEvents[IPCEvents["SHARDRESUME"] = 6] = "SHARDRESUME";
    IPCEvents[IPCEvents["SHARDDISCONNECT"] = 7] = "SHARDDISCONNECT";
    IPCEvents[IPCEvents["MASTEREVAL"] = 8] = "MASTEREVAL";
    IPCEvents[IPCEvents["RESTARTALL"] = 9] = "RESTARTALL";
    IPCEvents[IPCEvents["RESTART"] = 10] = "RESTART";
    IPCEvents[IPCEvents["FETCHUSER"] = 11] = "FETCHUSER";
    IPCEvents[IPCEvents["FETCHCHANNEL"] = 12] = "FETCHCHANNEL";
    IPCEvents[IPCEvents["FETCHGUILD"] = 13] = "FETCHGUILD";
})(IPCEvents = exports.IPCEvents || (exports.IPCEvents = {}));
var SharderEvents;
(function (SharderEvents) {
    SharderEvents["DEBUG"] = "debug";
    SharderEvents["MESSAGE"] = "message";
    SharderEvents["READY"] = "ready";
    SharderEvents["SPAWN"] = "spawn";
    SharderEvents["SHARD_READY"] = "shardReady";
    SharderEvents["SHARD_RECONNECT"] = "shardReconnect";
    SharderEvents["SHARD_RESUME"] = "shardResume";
    SharderEvents["SHARD_DISCONNECT"] = "shardDisconnect";
    SharderEvents["ERROR"] = "error";
})(SharderEvents = exports.SharderEvents || (exports.SharderEvents = {}));
