// kurasuta/dist/Sharding/ShardingManager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShardingManager = void 0;
const discord_js_1 = require("discord.js");
const MasterIPC_1 = require("../IPC/MasterIPC");
const Cluster_1 = require("../Cluster/Cluster");
const Constants_1 = require("../Util/Constants");
const events_1 = require("events");
const os_1 = require("os");
const cluster_1 = __importDefault(require("cluster"));
const Util = __importStar(require("../Util/Util"));
const node_fetch_1 = __importDefault(require("node-fetch"));
class ShardingManager extends events_1.EventEmitter {
    constructor(path, options) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        super();
        this.path = path;
        this.clusters = new Map();
        this.clusterCount = Number((_a = options.clusterCount) !== null && _a !== void 0 ? _a : (0, os_1.cpus)().length);
        this.guildsPerShard = Number((_b = options.guildsPerShard) !== null && _b !== void 0 ? _b : 1000);
        this.clientOptions = (_c = options.clientOptions) !== null && _c !== void 0 ? _c : { intents: discord_js_1.Intents.FLAGS.GUILDS };
        this.development = (_d = options.development) !== null && _d !== void 0 ? _d : false;
        this.shardCount = options.shardCount ? Number(options.shardCount) : 'auto';
        this.client = (_e = options.client) !== null && _e !== void 0 ? _e : discord_js_1.Client;
        this.respawn = (_f = options.respawn) !== null && _f !== void 0 ? _f : true;
        this.ipcSocket = (_g = options.ipcSocket) !== null && _g !== void 0 ? _g : 9999;
        this.retry = (_h = options.retry) !== null && _h !== void 0 ? _h : true;
        this.timeout = Number((_j = options.timeout) !== null && _j !== void 0 ? _j : 30000);
        this.token = options.token;
        this.nodeArgs = options.nodeArgs;
        this.ipc = new MasterIPC_1.MasterIPC(this);
        this.ipc.on('debug', msg => this._debug(`[IPC] ${msg}`));
        this.ipc.on('error', err => this.emit(Constants_1.SharderEvents.ERROR, err));
        if (!this.path)
            throw new Error('You need to supply a Path!');
    }
    async spawn() {
        if (cluster_1.default.isPrimary) {
            if (this.shardCount === 'auto') {
                this._debug('Fetching Session Endpoint');
                const { shards: recommendShards } = await this._fetchSessionEndpoint();
                this.shardCount = Util.calcShards(recommendShards, this.guildsPerShard);
                this._debug(`Using recommend shard count of ${this.shardCount} shards with ${this.guildsPerShard} guilds per shard`);
            }
            this._debug(`Starting ${this.shardCount} Shards in ${this.clusterCount} Clusters!`);
            if (this.shardCount < this.clusterCount) {
                this.clusterCount = this.shardCount;
            }
            const shardArray = [...Array(this.shardCount).keys()];
            const shardTuple = Util.chunk(shardArray, this.clusterCount);
            const failed = [];
            if (this.nodeArgs)
                cluster_1.default.setupPrimary({ execArgv: this.nodeArgs });
            for (let index = 0; index < this.clusterCount; index++) {
                const shards = shardTuple.shift();
                const cluster = new Cluster_1.Cluster({ id: index, shards, manager: this });
                this.clusters.set(index, cluster);
                try {
                    await cluster.spawn();
                }
                catch (_a) {
                    this._debug(`Cluster ${cluster.id} failed to start`);
                    this.emit(Constants_1.SharderEvents.ERROR, new Error(`Cluster ${cluster.id} failed to start`));
                    if (this.retry) {
                        this._debug(`Requeuing Cluster ${cluster.id} to be spawned`);
                        failed.push(cluster);
                    }
                }
            }
            if (this.retry)
                await this.retryFailed(failed);
        }
        else {
            return Util.startCluster(this);
        }
    }
    async restartAll() {
        this._debug('Restarting all Clusters!');
        for (const cluster of this.clusters.values()) {
            await cluster.respawn();
        }
    }
    async restart(clusterID) {
        const cluster = this.clusters.get(clusterID);
        if (!cluster)
            throw new Error('No Cluster with that ID found.');
        this._debug(`Restarting Cluster ${clusterID}`);
        await cluster.respawn();
    }
    fetchClientValues(prop) {
        return this.ipc.broadcast(`this.${prop}`);
    }
    eval(script) {
        return new Promise((resolve, reject) => {
            try {
                // tslint:disable-next-line:no-eval
                return resolve(eval(script));
            }
            catch (error) {
                reject(error);
            }
        });
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    once(event, listener) {
        return super.once(event, listener);
    }
    async retryFailed(clusters) {
        const failed = [];
        for (const cluster of clusters) {
            try {
                this._debug(`Respawning Cluster ${cluster.id}`);
                await cluster.respawn();
            }
            catch (_a) {
                this._debug(`Cluster ${cluster.id} failed, requeuing...`);
                failed.push(cluster);
            }
        }
        if (failed.length) {
            this._debug(`${failed.length} Clusters still failed, retry...`);
            return this.retryFailed(failed);
        }
    }
    async _fetchSessionEndpoint() {
        if (!this.token)
            throw new Error('No token was provided!');
        const res = await (0, node_fetch_1.default)(`${Constants_1.http.api}/v${Constants_1.http.version}/gateway/bot`, {
            method: 'GET',
            headers: { authorization: `Bot ${this.token.replace(/^Bot\s*/i, '')}` }
        });
        if (res.ok)
            return res.json();
        throw new Error(`Invalid Session Endpoint response: ${res.status} ${res.statusText}`);
    }
    _debug(message) {
        this.emit(Constants_1.SharderEvents.DEBUG, message);
    }
}
exports.ShardingManager = ShardingManager;
