// veza/dist/lib/Structures/NodeMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeMessage = void 0;
const binarytf_1 = require("binarytf");
const Header_1 = require("../Util/Header");
class NodeMessage {
    /**
     * @since 0.0.1
     * @param client The socket that received this message.
     * @param id The ID of the message.
     * @param receptive Whether or not this message accepts a reply.
     * @param data The data received from the socket.
     */
    constructor(client, id, receptive, data) {
        /**
         * The client that received this message.
         * @since 0.0.1
         */
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The id of this message.
         * @since 0.0.1
         */
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The data received from the socket.
         * @since 0.0.1
         */
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Whether the message is receptive or not.
         * @since 0.0.1
         */
        Object.defineProperty(this, "receptive", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperties(this, {
            client: { value: client },
            id: { value: id, enumerable: true },
            receptive: { value: receptive, enumerable: true }
        });
        this.data = data;
    }
    /**
     * Reply to the socket.
     * @since 0.0.1
     * @param content The content to send.
     */
    reply(content) {
        if (this.receptive && this.client.socket.writable) {
            this.client.socket.write((0, Header_1.createFromID)(this.id, false, (0, binarytf_1.serialize)(content)));
        }
    }
    /**
     * Freeze the message.
     * @since 0.0.1
     */
    freeze() {
        return Object.freeze(this);
    }
    /**
     * The toJSON overload for JSON.stringify.
     * @since 0.0.1
     */
    toJSON() {
        return {
            id: this.id,
            data: this.data,
            receptive: this.receptive
        };
    }
    /**
     * The toString overload for string casting.
     * @since 0.0.1
     */
    toString() {
        return `NodeMessage<${this.id}>`;
    }
}
exports.NodeMessage = NodeMessage;
